jenkins/aws-janitor/main.go

```diff
@@ -0,0 +1,786 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/golang/glog"
+)
+
+var regionsRaw = flag.String("regions", "us-west-2", "Comma separated list of regions to query for resources")
+var maxTTL = flag.Duration("ttl", 24*time.Hour, "Maximum time before we attempt deletion of a resource. Set to 0s to nuke all non-default resources.")
+var path = flag.String("path", "", "S3 path to store mark data in (required)")
+
+type awsResourceType interface {
+	// MarkAndSweep queries the resource in a specific region, using
+	// the provided session (which has account-number acct), calling
+	// res.Mark(<resource>) on each resource and deleting
+	// appropriately.
+	MarkAndSweep(sess *session.Session, acct string, region string, res *awsResourceSet) error
+}
+
+// AWS resource types known to this script, in dependency order.
+var awsResourceTypes = []awsResourceType{
+	autoScalingGroups{},
+	launchConfigurations{},
+	instances{},
+	// Addresses
+	// NetworkInterfaces
+	subnets{},
+	securityGroups{},
+	// NetworkACLs
+	// VPN Connections
+	internetGateways{},
+	routeTables{},
+	vpcs{},
+	dhcpOptions{},
+	volumes{},
+}
+
+type awsResource interface {
+	// ARN returns the AWS ARN for the resource
+	// (c.f. http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html). This
+	// is only used for uniqueness in the Mark set, but ARNs are
+	// intended to be globally unique across regions and accounts, so
+	// that works.
+	ARN() string
+}
+
+// awsResourceSet keeps track of the first time we saw a particular
+// ARN, and the global TTL. See Mark() for more details.
+type awsResourceSet struct {
+	firstSeen map[string]time.Time // ARN -> first time we saw
+	marked    map[string]bool      // ARN -> seen this run
+	ttl       time.Duration
+}
+
+func LoadResourceSet(sess *session.Session, p *s3path, ttl time.Duration) (*awsResourceSet, error) {
+	s := &awsResourceSet{firstSeen: make(map[string]time.Time), marked: make(map[string]bool), ttl: ttl}
+	svc := s3.New(sess, &aws.Config{Region: aws.String(p.region)})
+	resp, err := svc.GetObject(&s3.GetObjectInput{Bucket: aws.String(p.bucket), Key: aws.String(p.key)})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NoSuchKey" {
+			return s, nil
+		}
+		return nil, err
+	}
+	defer resp.Body.Close()
+	json.NewDecoder(resp.Body).Decode(&s.firstSeen)
```

Comment:
I went ahead and made this a hard error. If we somehow saved something that we can't actually restore, there's either a config issue (pointing to the wrong file) or a bug on save, and in either case we should avoid rewriting it, probably.